### ISSUE TITLE
`bundle exec <cmd>` gives "command not found" on JRuby+Windows

### ISSUE THREAD

**Issue by eregon** [jruby/jruby#6053]
**Environment Information**
  
Provide at least:
* JRuby 9.2.9.0
* Windows

See https://github.com/rubygems/bundler/issues/7602 and https://github.com/ruby/setup-ruby/issues/18#issuecomment-583777248

It's unclear if this is a Bundler bug or JRuby bug, but given that it works on MRI on Windows, and I it seems unlikely Bundler has code specific to MRI for `bundle exec`, I think it would be a JRuby bug.

Note that if there is any argument, the bug does not happen:
```
$ bundle exec rake
bundler: command not found: rake
Install missing gem executables with `bundle install`

$ bundle exec rake default
OK
```

Therefore a workaround is to replace `bundle exec rake` with `bundle exec rake default` but it's obviously just a workaround.

**Comment by slonopotamus** [comment]
This was fixed on Bundler side: https://github.com/rubygems/rubygems/issues/3256#issuecomment-691948644

**Comment by eregon** [comment]
Let's close then, thanks for checking.
@slonopotamus Could you file a separate issue for [the warning](https://github.com/rubygems/rubygems/issues/3256#issuecomment-691948644) on this tracker?

**Comment by eregon** [comment]
This seems to be an issue again:
https://github.com/eregon/childprocess/runs/2869731199?check_suite_focus=true
```
bundler: command not found: rake
Install missing gem executables with `bundle install`
```
That's JRuby 9.2.19.0 and Bundler 2.2.20.

**Comment by headius** [comment]
Unable to reproduce with a clean download of JRuby 9.2.19.0, Bundler 2.2.21, Rake 12.3.3:

```
C:\jruby-9.2.19.0\tmp>jruby -S bundle install
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using rake 12.3.3
Using bundler 2.2.21
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.

C:\jruby-9.2.19.0\tmp>jruby -S bundle exec rake -T
rake aborted!
No Rakefile found (looking for: rakefile, Rakefile, rakefile.rb, Rakefile.rb)
C:/jruby-9.2.19.0/bin/bundle:23:in `<main>'
(See full trace by running task with --trace)
```

**Comment by headius** [comment]
Since I cannot reproduce this with the standard JRuby 9.2.19.0 zip download, it seems the problem must lie somewhere else. 

If you are able to reproduce in isolation from `setup-ruby` please open a new issue, since this one was resolved by a Bundler update. Any future issue found to be JRuby's fault should be a new bug report.

**Comment by eregon** [comment]
> `jruby -S bundle exec rake -T`

The bug only happens with no arguments at all, as mentioned in the description, could you try literally `bundle exec rake` with `jruby` in PATH to see if that reproduces?

**Comment by headius** [comment]
```
C:\jruby-9.2.19.0\tmp>jruby -S bundle exec rake
rake aborted!
No Rakefile found (looking for: rakefile, Rakefile, rakefile.rb, Rakefile.rb)
C:/jruby-9.2.19.0/bin/bundle:23:in `<main>'
(See full trace by running task with --trace)
```

**Comment by headius** [comment]
Also

```
C:\jruby-9.2.19.0\tmp>bundle exec rake
rake aborted!
No Rakefile found (looking for: rakefile, Rakefile, rakefile.rb, Rakefile.rb)
C:\jruby-9.2.19.0\bin\bundle:23:in `<main>'
(See full trace by running task with --trace)
```

**Comment by eregon** [comment]
What if there is actually a `Rakefile`? (say, an empty one)
The error might happen only after.

There is nothing special to ruby/setup-ruby for JRuby on Windows, so it's a bit surprising it does not repro.

**Comment by slonopotamus** [comment]
What JVM version do you both use?

**Comment by headius** [comment]
```
C:\jruby-9.2.19.0\tmp>type Rakefile
task :default do
  puts "rake ran ok"
end
C:\jruby-9.2.19.0\tmp>bundle exec rake
rake ran ok

C:\jruby-9.2.19.0\tmp>jruby -v
jruby 9.2.19.0 (2.5.8) 2021-06-15 55810c552b OpenJDK 64-Bit Server VM 11.0.10+9 on 11.0.10+9 +jit [mswin32-x86_64]
```

**Comment by headius** [comment]
FWIW I confirmed it also works with the tarball, which appears to be how setup-ruby installs JRuby on Windows.

```
C:\tmp>tar xzf "C:\Users\Charles Oliver Nutte\Downloads\jruby-dist-9.2.19.0-bin.tar.gz"

C:\tmp>cd jruby-9.2.19.0

C:\tmp\jruby-9.2.19.0>set PATH=%PATH%;C:\tmp\jruby-9.2.19.0\bin

C:\tmp\jruby-9.2.19.0>mkdir tmp

C:\tmp\jruby-9.2.19.0>copy ..\..\jruby-9.2.19.0\tmp\Rakefile ..\..\jruby-9.2.19.0\tmp\Gemfile tmp
        2 file(s) copied.

C:\tmp\jruby-9.2.19.0>cd tmp

C:\tmp\jruby-9.2.19.0\tmp>gem install bundler
Fetching bundler-2.2.21.gem
Successfully installed bundler-2.2.21
Parsing documentation for bundler-2.2.21
Terminate batch job (Y/N)? y

C:\tmp\jruby-9.2.19.0\tmp>bundle install
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using rake 12.3.3
Using bundler 2.2.21
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.

C:\tmp\jruby-9.2.19.0\tmp>bundle exec rake
rake ran ok
```

**Comment by deivid-rodriguez** [comment]
I had a brief look at this today too, and I could confirm that the issue exists as reported. My guess is a bug in rubygems. Apparently `bundle install` is overwriting the `rake` exectutable with a shebang like this: `#! jruby`, which can't be later launched by bundler.

@headius I think you're not reproducing the issue because `rake` is already globally installed (See the "Using..." output in your `bundle install` command), so bundler is not installing anything and the binstub is not being overwritten with the bad content mentioned above. The difference with `ruby/setup-ruby` I believe is that it installs by default to `vendor/bundler`, so `rake` is not there and the bad binstub gets written.

**Comment by headius** [comment]
@deivid-rodriguez Just to confirm, there does not appear to be a bug in JRuby causing this, right?

If there turns out to be something we need to fix, please open a new issue or pull request. The original issue here was resolved by a bundler update.

**Comment by deivid-rodriguez** [comment]
That's my current guess, yeah. The current state of things where this issue is closed but the one in rubygems is opened seems accurate. I just posted my comment here just to clarify that the issue _is_ reproducible.

**Comment by headius** [comment]
@deivid-rodriguez Thank you! Good to know that this is not our bug (again). Let me know if there's anything I can do to help.

**Comment by deivid-rodriguez** [comment]
Well, it turns out this might be a bug in jruby after all.

Rubygems installs `.bat` wrappers for gem executables on Windows to compensate for the lack of support for shebangs. They look like this:

```
@ECHO OFF
@"%~dp0ruby.exe" "%~dpn0" %*
```

On CRuby + Windows, `Kernel.exec` properly detects and uses these scripts:

```

PS V:\> ruby -e "Kernel.exec 'rake'"
-e:1:in `exec': No such file or directory - rake (Errno::ENOENT)
        from -e:1:in `<main>'

PS V:\> bundle
Fetching gem metadata from https://rubygems.org/
Fetching gem metadata from https://rubygems.org/.
Fetching rake 13.0.3
Installing rake 13.0.3
Using bundler 2.3.0.dev
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.

PS V:\> ruby -e "Kernel.exec 'rake'"
rake aborted!
No Rakefile found (looking for: rakefile, Rakefile, rakefile.rb, Rakefile.rb)

(See full trace by running task with --trace)

PS V:\> rm C:\Ruby30-x64\bin\rake.bat

PS V:\> ruby -e "Kernel.exec 'rake'"
-e:1:in `exec': No such file or directory - rake (Errno::ENOENT)
        from -e:1:in `<main>'
```

However, on JRruby + Windows, they seem to be ignored:

```
PS V:\> jruby -e "Kernel.exec 'rake'"
Errno::ENOENT: No such file or directory - rake
  _exec_internal at org/jruby/RubyKernel.java:1767
            exec at uri:classloader:/jruby/kernel/kernel.rb:4
          <main> at -e:1

PS V:\> bundle
Fetching gem metadata from https://rubygems.org/
Fetching gem metadata from https://rubygems.org/.
Fetching rake 13.0.3
Installing rake 13.0.3
Using bundler 2.3.0.dev
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.

PS V:\> jruby -e "Kernel.exec 'rake'"
Errno::ENOENT: No such file or directory - No message available
  _exec_internal at org/jruby/RubyKernel.java:1767
            exec at uri:classloader:/jruby/kernel/kernel.rb:4
          <main> at -e:1
```

**Comment by deivid-rodriguez** [comment]
Since my current guess is that this is a bug in JRuby, and @headius told me to open a new issue in that case, I did that: #6745.
